Adding wrappers for `pod.babashka.buddy.sign.jwt`

brandonstubbs opened the pull request:
Started adding wrappers for `pod.babashka.buddy.sign.jwt`.

Before completing this, I wanted to discuss some ideas on the best way we could load the private keys which uses BouncyCastle under the hood.

For now, I have created a new `sign` function that takes extra opts in order to know how to load the key. Any other ideas or this is the best way to move ahead with this?

borkdude commented on .gitignore:
```diff
@@ -6,3 +6,7 @@ classes
 /pod-babashka-buddy.jar
 /pod-babashka-buddy.build_artifacts.txt
+.clj-kondo
+.lsp/.cache
```
I think you could just add

```
.cache
```
and this would be sufficient for both clojure-lsp and clj-kondo.

brandonstubbs replied:
Thanks for the info 👍

borkdude commented on src/pod/babashka/buddy.clj:
```diff
@@ -173,6 +184,7 @@
     (.toString baos "utf-8")))
 
 (defn -main [& _args]
+  (Security/addProvider (BouncyCastleProvider.))
```
I guess this isn't something we can do at compile time, outside of the -main right?

brandonstubbs replied:
One would think, however when adding it outside the `-main` the provider is not found.
I also tried adding it as an [AutomaticFeature](https://github.com/oracle/graal/issues/2800#issuecomment-1034705966) but didn't have much luck either.

borkdude replied:
Good to know.

borkdude commented:
There are a few things we could do here with the custom `sign` API.

1. Support your API. Question: do you expect the three options to be sufficient or do you expect more variations in the future?
2. Is this Key object argument in any way serializable, e.g. as a map or vector with bytes or so? Then we could define a transit handler back and forth for it and support the normal buddy API using some wrappers. (I can help with this)
3. If not 2, then we can still wrap `sign` and the `buddy.core.keys` functions like this: In the wrapper, we accept an argument, define the key object in some atom under an id and send back then id to the pod client. When invoking `sign`, our `sign` wrapper looks up the key in the atom by the id and then calls buddy's sign with the key.

If 2 and 3 sound too complicated and 1 is expected to be sufficient, I'm still ok with doing that, but it would be nice if we could keep things as close to the original as possible.

Perhaps @cap10morgan also has some ideas here.

brandonstubbs commented:
Thanks for the detailed feedback!

I have managed to find a way to tackle it as number 2.
I have added a new function for now `sign2`. Until all algos tested.

With these changes, we will only need to wrap the `buddy.core.keys` and return encoded keys as a `byte[]`. That will then need to be sent with the `jwt` sign/unsign functions. The encoded `byte[]` will need to be coerced back into a key.
It seems to be working with the RCA PEM key.

I will add the other key/algo types tomorrow to make sure it works for all, this looks like a way better way to keep as close as possible with the original.

borkdude commented:
@brandonstubbs Excellent! Any reason to not expose `sign2` as `sign`?

brandonstubbs commented:
@borkdude that's the plan, just kept both options available until I can confirm all algos work 😄 Then I will refactor the old one out, sign2 becomes sign

Oddsor commented:
@brandonstubbs I'm really looking forward to this one for a side project!
Since there's been no update in a while I assume you've been busy. If so I would love to help move this towards the finish line.

brandonstubbs commented:
@Oddsor sorry for the delay, I got married and was away on my honeymoon. I got back this week - I will aim to get this completed over the weekend

Oddsor commented:
> @Oddsor sorry for the delay, I got married and was away on my honeymoon. I got back this week - I will aim to get this completed over the weekend

Wow, congrats! 👏 And of course no worries, I just figured I would offer to contribute as I'm excited to give this a whirl soon 😊

brandonstubbs commented:
@Oddsor Thank you very much 😄 

I have pushed the changes for where I have got up to.
At the moment there seems to be an issue with the native build and BouncyCastle SecurityProvider. Will need some time to dig into it tomorrow, or if you have any ideas!

```
----- Error --------------------------------------------------------------------
Type:     clojure.lang.ExceptionInfo
Message:  Unable to create OpenSSL PBDKF: PBKDF-OpenSSL SecretKeyFactory not available
Data:     {:type "class org.bouncycastle.openssl.PEMException"}
Location: /Users/brandonstubbs/projects/clojure/pod-babashka-buddy/test.clj:63:18
```

brandonstubbs commented:
I have completed adding the wrappers.

@Oddsor Do you want to test it out and see if it works for you?

This currently will only work with GraalVM version 22.0.
GraalVM version 22.1 is throwing errors for which there is already an [issue](https://github.com/oracle/graal/issues/4419) open.

Following your checks, we can mark this ready for merge.

Oddsor commented:
> I have completed adding the wrappers.
> 
> @Oddsor Do you want to test it out and see if it works for you?
> 
> This currently will only work with GraalVM version 22.0. GraalVM version 22.1 is throwing errors for which there is already an [issue](https://github.com/oracle/graal/issues/4419) open.
> 
> Following your checks, we can mark this ready for merge.

Great work and thanks! 👏 

Apologies for the delayed response; I'm selling an apartment this week so most of my non-work time goes towards that up till Thursday 🙈 

However I did do a quick build of this and discovered that the exact use-case I was looking to test also depends on the `jwk->private-key` function from here: https://github.com/funcool/buddy-core/blob/master/src/buddy/core/keys.clj#L70

However, I think we can safely say that is outside the scope of this PR; I'll see if I can get some time in to implement the missing bits to cover my use-case 😅

brandonstubbs commented:
@Oddsor Sorry for the delay, just got some time to work on this again - I added most of the key functions. The one that you were missing is there now

brandonstubbs commented:
[build fails](https://app.circleci.com/pipelines/github/babashka/pod-babashka-buddy/34/workflows/4c95f35f-c0b6-4c21-93ba-3cd4f430f978/jobs/107)  for `linux-aarch64`.
It just seems to hang

```
========================================================================================================================
GraalVM Native Image: Generating 'pod-babashka-buddy'...
========================================================================================================================
Error: Image build request failed with exit status 1
com.oracle.svm.driver.NativeImage$NativeImageError: Image build request failed with exit status 1
	at com.oracle.svm.driver.NativeImage.showError(NativeImage.java:1770)
	at com.oracle.svm.driver.NativeImage.build(NativeImage.java:1477)
	at com.oracle.svm.driver.NativeImage.performBuild(NativeImage.java:1438)
	at com.oracle.svm.driver.NativeImage.main(NativeImage.java:1425)
```

As I can't trigger builds I don't want to spam commits to try debug
![image](https://user-images.githubusercontent.com/3731434/173719921-5dddf056-e45a-41c6-b9db-f5f8dc751739.png)

Oddsor commented:
> @Oddsor Sorry for the delay, just got some time to work on this again - I added most of the key functions. The one that you were missing is there now

Wow! I gave it a try and it works for my use-case (using a jwk-key) 🎉 

It took me an hour to figure out that there was some cached nonsense in my project folder that caused the pod to not work as expected 😅
But yeah, it works! Thanks for your awesome work 👏 🙏

borkdude commented:
So, good to merge now?

brandonstubbs commented:
@Oddsor no problem at all! 😄 Hope your apartment sale went smoothly!

@borkdude it is ready codewise, I am just unsure why the `linux-aarch64` build is [failing](https://app.circleci.com/pipelines/github/babashka/pod-babashka-buddy/34/workflows/4c95f35f-c0b6-4c21-93ba-3cd4f430f978/jobs/107) with minimal feedback

brandonstubbs commented:
I am going to spin up an arm machine my side to test

brandonstubbs commented:
It works on the machine I tested on

```
========================================================================================================================
GraalVM Native Image: Generating 'pod-babashka-buddy'...
========================================================================================================================
[1/7] Initializing...                                                                                    (5.0s @ 0.24GB)
The URL protocol jar is not tested and might not work as expected.
Supported URL protocols enabled by default: file,resource. Supported URL protocols available on demand: http,https.
--initialize-at-build-time without arguments has been deprecated when not using --diagnostics-mode. With GraalVM 22.0.0 --initialize-at-build-time will only work with --diagnostics-mode for debugging purposes.
The reason for deprecation is that --initalize-at-build-time does not compose, i.e., a single library can make assumptions that the whole classpath can be safely initialized at build time; that assumption is often incorrect.
 Version info: 'GraalVM 22.0.0.2 Java 11 CE'
 1 user-provided feature(s)
  - graalvm.feature.BouncyCastleFeature
[2/7] Performing analysis...  [**********]                                                              (39.2s @ 1.40GB)
   9,894 (87.96%) of 11,248 classes reachable
  11,780 (49.06%) of 24,009 fields reachable
  43,357 (59.23%) of 73,203 methods reachable
     875 classes,   398 fields, and 1,551 methods registered for reflection
      63 classes,    71 fields, and    53 methods registered for JNI access
[3/7] Building universe...                                                                               (2.7s @ 1.63GB)
[4/7] Parsing methods...      [**]                                                                       (2.5s @ 1.89GB)
[5/7] Inlining methods...     [****]                                                                     (3.9s @ 0.98GB)
[6/7] Compiling methods...    [****]                                                                    (21.0s @ 1.08GB)
[7/7] Creating image...                                                                                  (7.3s @ 1.63GB)
  19.77MB (36.29%) for code area:   30,599 compilation units
  25.67MB (47.12%) for image heap:   7,645 classes and 358,580 objects
   9.04MB (16.59%) for other data
  54.48MB in total
------------------------------------------------------------------------------------------------------------------------
Top 10 packages in code area:                               Top 10 object types in image heap:
   2.33MB clojure                                              6.01MB byte[] for general heap data
   1.37MB clojure.lang                                         4.01MB java.lang.Class
 779.63KB java.util                                            1.95MB java.lang.String
 752.80KB org.bouncycastle.math.ec.custom.sec                  1.76MB java.lang.Object[]
 669.63KB java.lang.invoke                                     1.61MB byte[] for java.lang.String
 581.02KB com.oracle.svm.core.reflect                        966.48KB clojure.lang.PersistentHashMap$BitmapIndexedNode
 575.70KB org.bouncycastle.crypto.engines                    540.89KB java.util.HashMap$Node
 557.99KB clojure.spec                                       522.94KB java.util.concurrent.ConcurrentHashMap$Node
 462.80KB com.sun.crypto.provider                            513.57KB clojure.lang.Symbol
 368.12KB sun.security.x509                                  402.09KB java.lang.String[]
      ... 315 additional packages                                 ... 3711 additional object types
                                           (use GraalVM Dashboard to see all)
------------------------------------------------------------------------------------------------------------------------
                        8.8s (10.1% of total time) in 96 GCs | Peak RSS: 3.33GB | CPU load: 5.72
------------------------------------------------------------------------------------------------------------------------
Produced artifacts:
 /home/ubuntu/pod-babashka-buddy/pod-babashka-buddy (executable)
 /home/ubuntu/pod-babashka-buddy/pod-babashka-buddy.build_artifacts.txt
========================================================================================================================
```

```
$ lsb_release -a
No LSB modules are available.
Distributor ID:	Ubuntu
Description:	Ubuntu 22.04 LTS
Release:	22.04
Codename:	jammy

$ uname -a
Linux ip-10-0-101-87 5.15.0-1009-aws #11-Ubuntu SMP Thu May 26 19:39:49 UTC 2022 aarch64 aarch64 aarch64 GNU/Linux

$ java -version
openjdk version "11.0.15" 2022-04-19
OpenJDK Runtime Environment (build 11.0.15+10-Ubuntu-0ubuntu0.22.04.1)
OpenJDK 64-Bit Server VM (build 11.0.15+10-Ubuntu-0ubuntu0.22.04.1, mixed mode)
```

borkdude commented:
Strange!

borkdude commented:
Does upgrading to CI GraalVM 22.1.0 help maybe?

brandonstubbs commented:
Unfortunately no, it is blocked by [this issue](https://github.com/oracle/graal/issues/4419) adding the BouncyCastle Feature it appears in that version. I will ping them on slack and see if anyone has any pointers

borkdude commented:
OK, I'd say we want on that then

borkdude commented:
@brandonstubbs Looks good. So what was the solution since last time?

borkdude commented:
I'd appreciate if you didn't force-push, then it's easier to see the diffs.

borkdude commented on deps.edn:
```diff
@@ -7,7 +7,8 @@
                     :extra-deps {org.clojure/clojure {:mvn/version "1.10.3"}
-                                 org.graalvm.nativeimage/svm {:mvn/version "22.0.0.2"}}}
+                                 ;; NOTE: This needs to match the GRAALVM Version
+                                 org.graalvm.nativeimage/svm {:mvn/version "22.1.0"}}}
```
I wonder why we ever added this. I don't think it's necessary.

brandonstubbs replied:
This is required as I add BouncyCastle as a [Feature](https://github.com/babashka/pod-babashka-buddy/pull/4/files#diff-f47055c890e398ca8c9c358ea148662bc5a2d2e3ab86ea33c1f5c43e46c5282c)

brandonstubbs commented:
@borkdude
Updated to GraalVM 22.1.0, but that wasn't the issue. It seemed that on `aarch-64` the build was hitting a deadlock - I updated all build images and tooling to the latest and it seemed to solve the issue.

Noted on the force push, that was to test something on the build and it didn't solve the issue so I just wanted to remove that commit.


This [build failure](https://app.circleci.com/pipelines/github/babashka/pod-babashka-buddy/36/workflows/a132e96b-903a-40bc-8fab-98d0c668cde5/jobs/115) is what finally showed there was a deadlock issue

borkdude commented:
@brandonstubbs Makes sense!

I think if we compile the uberjar using GraalVM bin/java itself, then we should not have to use the svm jar in the project. Adding the SVM jar in the project can give issues when it doesn't match, when upgrading graalvm. 

So here: https://github.com/babashka/pod-babashka-buddy/blob/120f664d4fe2b3e6bac85387355d487b776b91b1/script/compile#L10

we need to add these two lines:

https://github.com/babashka/babashka/blob/master/script/compile#L26-L27

and then it probably works without it.

brandonstubbs commented:
Ah, thanks @borkdude you are right! Made that change

borkdude commented:
@brandonstubbs Thanks. I'll make a new release, probably on Wednesday.
